internal/metrics/http_middleware_test.go

```diff
@@ -0,0 +1,162 @@
+				r := httptest.NewRequest(http.MethodGet, "/", nil)
+
+				return r
+			},
+			expected: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result := isSpdyRequest(tt.newRequestFn())
+			assert.Equal(t, tt.expected, result)
+		})
+	}
+}
+
+func TestHTTPMetricsMiddleware(t *testing.T) {
+	testCases := []struct {
+		name                string
+		setupRequest        func(*http.Request)
+		expectedRequestType string
+	}{
+		{
+			name:                "HTTP Request",
+			setupRequest:        func(_ *http.Request) {}, // No special headers for regular HTTP
+			expectedRequestType: "http",
+		},
+		{
+			name: "WebSocket Request",
+			setupRequest: func(req *http.Request) {
+				req.Header.Set("Upgrade", "websocket")
+				req.Header.Set("Connection", "upgrade")
+			},
+			expectedRequestType: "websocket",
+		},
+		{
+			name: "SPDY Request",
+			setupRequest: func(req *http.Request) {
+				req.Header.Set("Upgrade", "spdy/3.1")
+				req.Header.Set("Connection", "upgrade")
+			},
+			expectedRequestType: "spdy",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			testRegistry := prometheus.NewRegistry()
+
+			server := httptest.NewServer(HTTPMetricsMiddleware(
+				HTTPMiddlewareConfig{
+					Registry: testRegistry,
+					Next: http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+						w.WriteHeader(http.StatusOK)
+					}),
+				},
+			))
+			defer server.Close()
+
+			req, err := http.NewRequest(http.MethodGet, server.URL, nil)
+			require.NoError(t, err, "failed to create request")
+
+			tc.setupRequest(req)
+
+			resp, err := http.DefaultClient.Do(req)
+			require.NoError(t, err, "failed to send request")
+			defer resp.Body.Close()
+
+			metricFamilies, err := testRegistry.Gather()
+			require.NoError(t, err)
+
+			registeredLabels := extractLabelsFromMetrics(metricFamilies)
+			expectedLabels := map[string]map[string]string{
+				"twingate_gateway_http_requests_total": {
+					"type":   tc.expectedRequestType,
+					"method": "get",
+					"code":   "200",
+				},
+				"twingate_gateway_http_request_duration_seconds": {
+					"type":   tc.expectedRequestType,
+					"method": "get",
+					"code":   "200",
+				},
+				"twingate_gateway_http_request_size_bytes": {
+					"type":   tc.expectedRequestType,
+					"method": "get",
+					"code":   "200",
+				},
+				"twingate_gateway_http_response_size_bytes": {
+					"type":   tc.expectedRequestType,
+					"method": "get",
+					"code":   "200",
+				},
+			}
+			assert.Equal(t, expectedLabels, registeredLabels)
+		})
+	}
+}
+
+func extractLabelsFromMetrics(metricFamilies []*dto.MetricFamily) map[string]map[string]string {
+	registeredLabels := make(map[string]map[string]string, len(metricFamilies))
+
+	for _, mf := range metricFamilies {
+		labels := mf.GetMetric()[0].GetLabel()
+		for _, label := range labels {
+			value, ok := registeredLabels[mf.GetName()]
+			if !ok {
+				value = make(map[string]string)
+			}
```

Comment:
It should be simpler to create the map before looping the label e.g.
```go
func extractLabelsByMetric(metricFamilies []*dto.MetricFamily) map[string]map[string]string {
	labelsByMetric := make(map[string]map[string]string, len(metricFamilies))

	for _, family := range metricFamilies {
		labels := make(map[string]string)
		for _, label := range family.GetMetric()[0].GetLabel() {
			labels[label.GetName()] = label.GetValue()
		}

		labelsByMetric[family.GetName()] = labels
	}

	return labelsByMetric
}
```

Reply:
Updated in [bb9e635](https://github.com/Twingate/kubernetes-access-gateway/pull/49/commits/bb9e635ba9425a26136d54e771de9deda4a12da6)